File: backend/src/dearth/ModFilterService.ts
import { FileExtractor } from "./utils/FileExtractor.js";
import { FileOperator } from "./utils/FileOperator.js";
import { runFilterStrategies } from "./utils/StrategyRunner.js";
import { IFilterConfig } from "./types.js";
import { logger } from "../utils/logger.js";
import { MessageWS } from "../utils/socketio.js";

export class ModFilterService {
  private readonly extractor: FileExtractor;
  private readonly operator: FileOperator;
  private readonly config: IFilterConfig;
  private messageWS?: MessageWS;

  constructor(modsPath: string, movePath: string, config: IFilterConfig, messageWS?: MessageWS) {
    this.extractor = new FileExtractor(modsPath);
    this.operator = new FileOperator(movePath);
    this.config = config;
    this.messageWS = messageWS;
  }

  async filter(): Promise<void> {
    logger.info("开始模组筛选流程");
    const startTime = Date.now();

    try {
      const files = await this.extractor.extractFilesInfo();

      if (this.messageWS) {
        this.messageWS.filterModsStart(files.length);
      }

      const clientMods = await this.identifyClientSideMods(files);
      const result = await this.operator.moveClientSideMods(clientMods);

      const duration = Date.now() - startTime;

      if (this.messageWS) {
        this.messageWS.filterModsComplete(clientMods.length, result.success, duration);
      }

      logger.info("模组筛选流程完成", { 
        识别到的客户端模组: clientMods.length,
        成功移动: result.success,
        跳过: result.skipped,
        失败: result.error
      });
    } catch (error) {
      if (this.messageWS) {
        this.messageWS.filterModsError(error instanceof Error ? error.message : String(error));
      }
      throw error;
    }
  }

  private async identifyClientSideMods(files: Array<{ filename: string; hash: string; mixins: any[]; infos: any[] }>): Promise<string[]> {
    return runFilterStrategies(files, this.config, this.messageWS);
  }
}

File: backend/dearth/modfilterservice.go
package dearth

import (
	"log/slog"
	"fmt"

	"dex/backend/dearth/types"
)

// ModFilterService is the main entry point for mod filtering.
type ModFilterService struct {
	extractor *FileExtractor
	operator  *FileOperator
	config    types.FilterConfig
}

func NewModFilterService(modsPath, movePath string, config types.FilterConfig) *ModFilterService {
	return &ModFilterService{
		extractor: NewFileExtractor(modsPath),
		operator:  NewFileOperator(movePath),
		config:    config,
	}
}

// Filter runs the full mod filtering workflow:
// 1. Extract file info from all jars
// 2. Run filter strategies to identify client-side mods
// 3. Move client-side mods to the .clientmod directory
func (mfs *ModFilterService) Filter() error {
	slog.Info("Starting mod filter workflow")

	files, err := mfs.extractor.ExtractFilesInfo()
	if err != nil {
		return fmt.Errorf("failed to extract file info: %w", err)
	}
	if len(files) == 0 {
		slog.Info("No jar files found, skipping filter")
		return nil
	}

	clientMods, err := RunFilterStrategies(files, mfs.config)
	if err != nil {
		return fmt.Errorf("failed to identify client-side mods: %w", err)
	}
	if len(clientMods) == 0 {
		slog.Info("No client-side mods identified")
		return nil
	}

	slog.Info("Identified client-side mods", "count", len(clientMods))

	result := mfs.operator.MoveClientSideMods(clientMods)
		slog.Info("Mod filter complete", "moved", result.Success, "skipped", result.Skipped, "errors", result.Error)

	return nil
}

// IdentifyOnly runs the filter strategies without moving files (preview/check-only).
func (mfs *ModFilterService) IdentifyOnly() ([]string, error) {
	files, err := mfs.extractor.ExtractFilesInfo()
	if err != nil {
		return nil, fmt.Errorf("failed to extract file info: %w", err)
	}
	return RunFilterStrategies(files, mfs.config)
}
